rapids-cmake/cpm/detail/gpv_write_file.cmake

```diff
@@ -0,0 +1,252 @@
+rapids_cpm_gpv_pretty_format_json
+---------------------------------
+
+
+.. versionadded:: v24.04.00
+
+Formats provided JSON to be easily read by humans
+#]=======================================================================]
+function(rapids_cpm_gpv_pretty_format_json _rapids_json_var)
+  set(pretty_json)
+  #[=[
+  - parse each line
+  - trim all leading spaces
+  - exclusive if line contains "}" or "]" decement indenter
+  - print with given indenter
+  - exclusive if line contains "{" or "[" increment indenter
+  ]=]
+  #
+  set(indent "  ")
+  set(indent_len 2)
+  set(current_indent "")
+  string(REPLACE "\n" ";" input "${${_rapids_json_var}}")
+
+  # Needed due to rules around cmake language list expansion:
+  #
+  # during evaluation of an Unquoted Argument. In such contexts, a string is divided into list
+  # elements by splitting on ; characters not following an unequal number of [ and ] characters
+  #
+  string(REPLACE "[" "``" input "${input}")
+  string(REPLACE "]" "~~" input "${input}")
+  foreach(line ${input})
+    string(STRIP "${line}" line)
+    set(push_indent)
+    set(pop_indent)
+    if(line MATCHES "(\\{|``)")
+      set(push_indent TRUE)
+    endif()
+    if(line MATCHES "(\\}|~~)")
+      set(pop_indent TRUE)
+    endif()
+
+    if(pop_indent AND NOT push_indent)
+      string(SUBSTRING "${current_indent}" ${indent_len} -1 current_indent)
+    endif()
+
+    string(APPEND pretty_json "${current_indent}${line}\n")
+
+    if(push_indent AND NOT pop_indent)
+      string(APPEND current_indent "${indent}")
+    endif()
+  endforeach()
+  string(REPLACE "``" "[" pretty_json "${pretty_json}")
+  string(REPLACE "~~" "]" pretty_json "${pretty_json}")
+  set(${_rapids_json_var} "${pretty_json}" PARENT_SCOPE)
+endfunction()
+
+#[=======================================================================[.rst:
+rapids_cpm_gpv_write_file
+-------------------------
+
+.. versionadded:: v24.04.00
+
+This function will write out the pinned version info to the provided files when we are in the root
+CMakeLists.txt
+#]=======================================================================]
+function(rapids_cpm_gpv_write_file)
+  if(NOT CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
+    return()
+  endif()
+
+  find_package(Git QUIET)
+
+  set(_rapids_json
+      [=[
+{
+"packages" : {
+]=])
+  foreach(package IN LISTS CPM_PACKAGES)
+    # Clear variables so we don't re-use them between packages when one package doesn't have a git
+    # url or sha
+    set(git_url)
+    set(git_sha)
+    rapids_cpm_gpv_extract_source_git_info(${package} git_url git_sha)
+    rapids_cpm_gpv_add_json_entry(_rapids_entry ${package} git_url git_sha)
+    string(APPEND _rapids_json "${_rapids_entry}")
+  endforeach()
+
+  # To make everything easier we add a fake package to the end so we don't have any trailing ','
+  set(post_amble
+      [=[
+  "trailing_place_holder" : {}
```

Comment:
This feels like a hack. Is this written to disk?

Reply:
It is, since it makes `rapids_cpm_gpv_pretty_format_json` significantly easier to implement.
We could remove it but would need to do another string removal pass IIRC

Reply:
Updated the implementation to not need the `trailing_place_holder` for packages.